### PR TITLE
feat(deadline): add X509 certificate expiration alarms

### PR DESCRIPTION
**Pull request by kozlove-aws**
### Problem

Certificates created in RFDK has expiration date but does not have procedure for rotation.
By default certificates will expire after 3 years, but there is still possibility that it expire and we need to have mechanism to warn customer that certificate have to be rotated.

### Solution

Was added a Lambda function for each certificate that checks secret with certificate every day and fill Metric with data how many days left before certificate will expire.
Alarm will be raised when certificate has only 15 days left to be valid.
Customer can provide an email and message about this alarm will be send in this case.

### Testing

Deployed example and validated:
- Lambda is created for each certificate
- Lambda is running every day
- Metric for secrets are created and values are filled
- Alarm is raised when value in metric became less than 15 days
- SNS topic is created and email about alarm was sent.
 
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts]:
```suggestion
  private addLambdaMonitorFunction(props: LambdaMonitorFunctionProps): LambdaFunction{
```

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts]: It seems a bit odd to have the code the monitor takes as a variable but then have the handler it runs as a hardcoded string. Let's follow the pattern that the class hierarchy is already using and pass this handler in as a `lambdaMonitorHandler` field from the concrete classes to the base class.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts]: Node 12 is coming up on EOL early next year. Maybe we should take the opportunity to upgrade this one and the actual Lambda to `Runtime.NODEJS_14_X`.

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/certificate.ts:93]: This should have unit testing

File: packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/index.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/index.ts]: Can we pass `process.env.UNIQUEID` into this function like you did for `getSecretsByTag`?

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts]: Can we share `metricName`, `namespace`, and `dimensionName` between this construct and the Lambda code where it also appears? I'm thinking they could be static constants in the Lambda code that we reference here.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:292]: The cost, security considerations, and resources deployed sections for both the X509CertificatePem and X509CertificatePkcs12 should be updated to include information about the lambda and monitor.

We might need to update the diagram for this as well.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts]: The comment says "data point hour" but then the Duration is `Duration.days(1)`. Is one of those incorrect? It looks like the Lambda that creates the metric runs daily, so maybe we just need to update the comment to say one data point per day.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts]: How come you went with a percentile here? Shouldn't we only be collecting one data point per day? Since the value of the metric is the number of days until expiry, maybe we should use `Minimum` here.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:361]: Did we figure out what this would cost to run every day indefinitely? It might make sense to run it weekly and set the threshold to a month, or make it configurable based on the `validFor` value. The `validFor` field defaults to 3 years, so we'd be fine to run weekly and notify with a month left in that case, but maybe if they set it for less than a year we could run daily and notify with 15 days left.

File: packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/index.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/index.ts]: Do we ever expect multiple certificates to be attached to a single tag? Looking at the code, it didn't look possible to me. If that's the case, maybe we should check if this list contains <= 1 items.

Thinking it through, if we switch the statistic from `p99` to `minimum` like I suggested in a previous comment, maybe we can leave this to iterate through the items. I think leaving it at `p99` will be a problem though, because if for some reason it did emit 2 metrics for 2 expiry dates, I think p99 would end up looking at the higher value.

**Review by horsmand — CHANGES_REQUESTED** [review]

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (kozlove-aws) [reply packages/aws-rfdk/lib/core/lib/x509-certificate.ts:361]: As I know Lambda price is extremely low. Running once  a day can guarantee that alarm will be raised in time even when customer replace this certificate manually.

File: packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/index.ts
Comment (kozlove-aws) [reply packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/index.ts]: It is possible if will Lambda generate more than one certificate but we do not expect it so we do not want to have any exceptions in this case. This code will work if we receive one ore more secrets with certificates to check.

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts]: Missing blank line

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/certificate.ts:106]: What's the default log level for a Lambda? Will these debug lines always appear or should we change them to `console.log`?

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts]: This comment doesn't really add any value beyond what the code already says, maybe we could change it to:
```suggestion
      // We're looking at the minimum value reported by the Lambda in a day, which should
      // be the only value, since it runs once daily.
```

File: packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/test/handler.test.ts
Comment (horsmand) [inline packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/test/handler.test.ts]: Missing blank line

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [reply packages/aws-rfdk/lib/core/lib/x509-certificate.ts]: Any reason not to upgrade the other one too?

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (horsmand) [reply packages/aws-rfdk/lib/core/lib/x509-certificate.ts:292]: I don't see this info added.

File: packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/index.ts
Comment (horsmand) [reply packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/index.ts]: If we do leave the logic like this where we are looping over the list, but don't expect it to have more than 1 item, maybe we should have a comment saying that we don't expect multiple items, but the monitor should still work if there is more than one since we're alarming on the minimum value.

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/certificate.ts
Comment (kozlove-aws) [reply packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/certificate.ts:106]: Added debug mode, switched off by default.

File: packages/aws-rfdk/lib/deadline/test/render-queue.test.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/deadline/test/render-queue.test.ts:2499]: I'm wondering if we want to create a Lambda function per certificate. In this test, we create a `RenderQueue` without providing any certificates which appears to create 4 certificates and therefor 4 Lambda functions and 4 IAM roles. This will bloat a CFN template and detract from its template [resource/file-size limits](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cloudformation-limits.html). Currently there is a limit of 500 resources per template and 51.2KB inline / 1MB S3 max CFN template size. I believe CDK uploads templates to the CDK bootstrap S3 bucket, so we are dealing with a 1MB template size limit.

I wonder if there's a way we could have a single Lambda function and CloudWatch Events rule per stack, aggregate the cert unique IDs and then have the lambda loop over them all.

As a CDK construct library, RFDK should be a good citizen and try to minimize what we contribute to our end-user's CFN template limits. What do you think?

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:359]: Can we change `ScheduleSecretMonitoring` &rarr; `ExpirationCheckRule`?

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:407]: we can probably just hard-code this in `X509CertificateBase.addLambdaMonitorFunction`, because right now its duplicated here and in `X509CertificatePkcs12`.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:385]: This change also deploys:

* IAM role
* IAM policy
* CloudWatch log group
* CloudWatch events rule
* SNS topic
* SNS subscription

according to a quick `cdk diff` I ran.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:188]: it might help for code readability to move these interfaces to be defined inline into the function definition where they're used:

```ts
private addLambdaMonitorFunction(props: {
    readonly uniqueValue: string;
    readonly lambdaLayers: ILayerVersion[];
    readonly lambdaCode: Code;
    readonly lambdaMonitorHandler: string;
    readonly tagCondition?: {
      [key: string]: any;
    };
  }): LambdaFunction{
```

this keeps the props definition closer to the code where it is used. What do you think?

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:331]: nit: can we rename this to `addExpirationDateTracking`?

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:342]: I think we can simplify this to:

```ts
    const certMetric = new Metric({
      metricName: CertificateRotationMonitor.METRIC_NAME,
      namespace: CertificateRotationMonitor.METRIC_NAMESPACE,
      dimensions: {
        [CertificateRotationMonitor.METRIC_DIMENSION]: props.uniqueValue,
      },
      // We're looking at the minimum value reported by the Lambda in a day, which should
      // be the only value, since it runs once daily.
      period: Duration.days(1),
      statistic: 'Minimum',
    });
```

File: packages/aws-rfdk/lib/core/test/x509-certificate.test.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/test/x509-certificate.test.ts:21]: it looks like we're missing tests for some of the new resources deployed by this code:

*   `AWS::SNS::Topic`
*   `AWS::IAM::Role`
*   `AWS::Events::Rule`
*   `Custom::LogRetention`

File: packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/handler.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/handler.ts:69]: This should be changed into an `async` function, since `cloudwatch.putMetricData` is asynchronous and we want it to complete before the Lambda function exits.

File: packages/aws-rfdk/lib/core/test/x509-certificate.test.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/test/x509-certificate.test.ts:403]: should we alarm on missing data? that means that our Lambda is not publishing metrics correctly, right?

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:353]: We should encrypt this SNS topic by default. It would be nice to be able to supply a KMS key in the props so we don't always create a KMS key per-certificate.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:350]: Can we add an `alarmDescription` here? I suspect this description will be included in the email notification. We should probably provide or link to the action that the recipient should take to rotate the certificates. See the following example in our example app for setting up similar metric/alarm/notification for EFS burst credits:

https://github.com/aws/aws-rfdk/blob/0933a9e5a569d0f03a266503bfdf59206a2ce41d/examples/deadline/All-In-AWS-Infrastructure-Basic/ts/lib/storage-tier.ts#L200-L285

File: packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/handler.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/handler.ts:55]: nit: AWS Certificate Manager uses the "Count" unit for their DaysToExpiry metric. Can we do the same?

File: packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/handler.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/lambdas/nodejs/cert-rotation-monitor/handler.ts:19]: Can we change this to "Certificate ARN"?

EDIT: Or perhaps "Certificate ID"? Is it possible to feed the cert ARN into this lambda instead? Because that might be easier for an end-user to reason about.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:289]: I tested this with an [`X509CertificatePkcs12`](https://docs.aws.amazon.com/rfdk/api/latest/docs/aws-rfdk.X509CertificatePkcs12.html) which derives from this base class, and found that it does not emit any metric. I'm not certain if this is because:
*   PKCS#12 certs don't have expiration
*   The command you are using to extract the certificate expiry does not work with PKCS#12 certificates
*   Something else...

That being said, based on the construct's API, PKCS#12 are always converted from another `X509CertificatePem` which will have its own metric/alarm. Depending on how RFDK instructs rotating these certs, we can probably do with out metrics/alarms on `X509CertificatePkcs12`. Should we move this logic into `X509CertificatePem` instead?

**Review by jusiskin — CHANGES_REQUESTED** [review]
Great solution to this problem. There are a few things we need to address before we can merge this.

**Pull request closed without merging**
